File: src/net/reichholf/dreamdroid/DatabaseHelper.java
/* © 2010 Stephan Reichholf <stephan at reichholf dot net>
 * 
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */

package net.reichholf.dreamdroid;

import java.util.ArrayList;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

/**
 * @author sre
 *
 */
public class DatabaseHelper extends SQLiteOpenHelper {
	public static final String KEY_ID = "_id";
	public static final String KEY_PROFILE = "profile";
	public static final String KEY_HOST = "host";
	public static final String KEY_STREAM_HOST = "streamhost";
	public static final String KEY_STREAM_PORT = "streamport";
	public static final String KEY_FILE_PORT = "fileport";
	public static final String KEY_PORT = "port";
	public static final String KEY_LOGIN = "login";
	public static final String KEY_USER = "user";
	public static final String KEY_PASS = "pass";
	public static final String KEY_SSL = "ssl";
	public static final String KEY_SIMPLE_REMOTE = "simpleremote";
	public static final String KEY_DEFAULT_REF = "default_ref";
	public static final String KEY_DEFAULT_REF_NAME = "default_ref_name";
	public static final String KEY_DEFAULT_REF_2 = "default_ref_2";
	public static final String KEY_DEFAULT_REF_2_NAME = "default_ref_2_name";

	public static final String DATABASE_NAME = "dreamdroid";
	private static final int DATABASE_VERSION = 7;
	private static final String PROFILES_TABLE_NAME = "profiles";

	private static final String PROFILES_TABLE_CREATE = "CREATE TABLE IF NOT EXISTS " + 
				PROFILES_TABLE_NAME + " (" +
				KEY_ID + " INTEGER PRIMARY KEY, " + 
				KEY_PROFILE + " TEXT, " +
				KEY_HOST + " TEXT, " +
				KEY_STREAM_HOST + " TEXT, " + 
				KEY_PORT + " INTEGER, " + 
				KEY_STREAM_PORT + " INTEGER, " + 
				KEY_FILE_PORT + " INTEGER, " +
				KEY_LOGIN + " BOOLEAN, " + 
				KEY_USER + " TEXT, " + 
				KEY_PASS + " TEXT, " + 
				KEY_SSL + " BOOLEAN, " + 
				KEY_SIMPLE_REMOTE + " BOOLEAN, " +
				KEY_DEFAULT_REF + " TEXT, " + 
				KEY_DEFAULT_REF_NAME + " TEXT, " +
				KEY_DEFAULT_REF_2 + " TEXT, " +
				KEY_DEFAULT_REF_2_NAME + " TEXT );";

	private static final String PROFILES_TABLE_UPGRADE_2_3 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD "
			+ KEY_STREAM_HOST + " TEXT;";

	private static final String PROFILES_TABLE_UPGRADE_3_4 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD "
			+ KEY_SIMPLE_REMOTE + " BOOLEAN;";

	private static final String PROFILES_TABLE_UPGRADE_4_5 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD "
			+ KEY_STREAM_PORT + " INTEGER;";
	
	private static final String PROFILES_TABLE_UPGRADE_5_6 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD "
			+ KEY_FILE_PORT + " INTEGER;";
	
	private static final String PROFILES_TABLE_UPGRADE_6_7_1 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD " + KEY_DEFAULT_REF + " TEXT;";
	private static final String PROFILES_TABLE_UPGRADE_6_7_2 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD " + KEY_DEFAULT_REF_NAME + " TEXT;";
	private static final String PROFILES_TABLE_UPGRADE_6_7_3 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD " + KEY_DEFAULT_REF_2 + " TEXT;";
	private static final String PROFILES_TABLE_UPGRADE_6_7_4 = "ALTER TABLE " + PROFILES_TABLE_NAME + " ADD " + KEY_DEFAULT_REF_2_NAME + " TEXT;";

	
	private Context mContext;
	/**
	 * @param context
	 * @param name
	 * @param factory
	 * @param version
	 */
	public DatabaseHelper(Context context) {
		super(context, DATABASE_NAME, null, DATABASE_VERSION);
		mContext = context;
	}

	/* (non-Javadoc)
	 * @see android.database.sqlite.SQLiteOpenHelper#onCreate(android.database.sqlite.SQLiteDatabase)
	 */
	@Override
	public void onCreate(SQLiteDatabase db) {
		db.execSQL(PROFILES_TABLE_CREATE);
	}
	
	private void upgrade6to7(SQLiteDatabase db){
		db.execSQL(PROFILES_TABLE_UPGRADE_6_7_1);
		db.execSQL(PROFILES_TABLE_UPGRADE_6_7_2);
		db.execSQL(PROFILES_TABLE_UPGRADE_6_7_3);
		db.execSQL(PROFILES_TABLE_UPGRADE_6_7_4);
	}
	
	/* (non-Javadoc)
	 * @see android.database.sqlite.SQLiteOpenHelper#onUpgrade(android.database.sqlite.SQLiteDatabase, int, int)
	 */
	@Override
	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
		if (oldVersion == 2) {
			db.execSQL(PROFILES_TABLE_UPGRADE_2_3);
			db.execSQL(PROFILES_TABLE_UPGRADE_3_4);
			db.execSQL(PROFILES_TABLE_UPGRADE_4_5);
			db.execSQL(PROFILES_TABLE_UPGRADE_5_6);
			upgrade6to7(db);
			db.setVersion(DATABASE_VERSION);
		} else if (oldVersion == 3) {
			db.execSQL(PROFILES_TABLE_UPGRADE_3_4);
			db.execSQL(PROFILES_TABLE_UPGRADE_4_5);
			db.execSQL(PROFILES_TABLE_UPGRADE_5_6);
			upgrade6to7(db);
			db.setVersion(DATABASE_VERSION);
		} else if (oldVersion == 4){
			db.execSQL(PROFILES_TABLE_UPGRADE_4_5);
			db.execSQL(PROFILES_TABLE_UPGRADE_5_6);
			upgrade6to7(db);
			db.setVersion(DATABASE_VERSION);
		} else if (oldVersion == 5){
			db.execSQL(PROFILES_TABLE_UPGRADE_5_6);
			upgrade6to7(db);
			db.setVersion(DATABASE_VERSION);
		} else if(oldVersion == 6){
			upgrade6to7(db);
			db.setVersion(DATABASE_VERSION);
		} else {
			db.execSQL("DROP TABLE IF EXISTS " + PROFILES_TABLE_NAME);
			db.setVersion(0);
		}
		if(oldVersion < newVersion)
			DreamDroid.scheduleBackup(mContext);
	}
	
	private ContentValues p2cv(Profile p){
		ContentValues values = new ContentValues();
		values.put(KEY_PROFILE, p.getName());
		values.put(KEY_HOST, p.getHost());
		values.put(KEY_STREAM_HOST, p.getStreamHostValue());
		values.put(KEY_PORT, p.getPort());
		values.put(KEY_STREAM_PORT, p.getStreamPort());
		values.put(KEY_FILE_PORT, p.getFilePort());
		values.put(KEY_LOGIN, p.isLogin());
		values.put(KEY_USER, p.getUser());
		values.put(KEY_PASS, p.getPass());
		values.put(KEY_SSL, p.isSsl());
		values.put(KEY_SIMPLE_REMOTE, p.isSimpleRemote());
		values.put(KEY_DEFAULT_REF, p.getDefaultRef());
		values.put(KEY_DEFAULT_REF_NAME, p.getDefaultRefName());
		values.put(KEY_DEFAULT_REF_2, p.getDefaultRef2());
		values.put(KEY_DEFAULT_REF_2_NAME, p.getDefaultRef2Name());
		return values;
	}
	
	/**
	 * @param p
	 */
	public boolean addProfile(Profile p) {
		SQLiteDatabase db = getWritableDatabase();
		if (db.insert(PROFILES_TABLE_NAME, null, p2cv(p)) > -1) {
			db.close();
			DreamDroid.scheduleBackup(mContext);
			return true;
		}
		db.close();
		return false;
	}

	/**
	 * @param p
	 */
	public boolean updateProfile(Profile p) {
		SQLiteDatabase db = getWritableDatabase();
		int numRows = db.update(PROFILES_TABLE_NAME, p2cv(p), KEY_ID + "=" + p.getId(), null);
		db.close();
		if (numRows == 1) {
			DreamDroid.scheduleBackup(mContext);
			DreamDroid.profileChanged(mContext, p);
			return true;
		}
		return false;
	}
	
	/**
	 * @param p
	 */
	public boolean deleteProfile(Profile p) {
		SQLiteDatabase db = getWritableDatabase();
		int numRows = db.delete(PROFILES_TABLE_NAME, KEY_ID + "=" + p.getId(), null);
		db.close();
		if (numRows == 1) {
			DreamDroid.scheduleBackup(mContext);
			return true;
		}

		return false;
	}

	/**
	 * @return Cursor for all Settings
	 */
	public ArrayList<Profile> getProfiles() {
		String[] columns = { KEY_ID, KEY_PROFILE, KEY_HOST, KEY_STREAM_HOST, KEY_PORT, KEY_STREAM_PORT, KEY_FILE_PORT, KEY_LOGIN, KEY_USER, KEY_PASS,
				KEY_SSL, KEY_SIMPLE_REMOTE, KEY_DEFAULT_REF, KEY_DEFAULT_REF_NAME, KEY_DEFAULT_REF_2, KEY_DEFAULT_REF_2_NAME };
		SQLiteDatabase db = getReadableDatabase();

		ArrayList<Profile> list = new ArrayList<Profile>();
		
		Cursor c = db.query(PROFILES_TABLE_NAME, columns, null, null, null, null, KEY_PROFILE);
		if(c.getCount() == 0){
			db.close();
			c.close();
			return list;
		}
		
		while(!c.isLast()){
			c.moveToNext();
			list.add(new Profile(c));
		}
		c.close();
		db.close();
		return list;
	}

	public Profile getProfile(int id) {
		String[] columns = { KEY_ID, KEY_PROFILE, KEY_HOST, KEY_STREAM_HOST, KEY_PORT, KEY_STREAM_PORT, KEY_FILE_PORT, KEY_LOGIN, KEY_USER, KEY_PASS,
				KEY_SSL, KEY_SIMPLE_REMOTE, KEY_DEFAULT_REF, KEY_DEFAULT_REF_NAME, KEY_DEFAULT_REF_2, KEY_DEFAULT_REF_2_NAME };
		SQLiteDatabase db = getReadableDatabase();
		Cursor c = db.query(PROFILES_TABLE_NAME, columns, KEY_ID + "=" + id, null, null, null, KEY_PROFILE);

		Profile p = new Profile();
		if(c.getCount() == 1 && c.moveToFirst())
			p.set(c);
		c.close();
		db.close();
		return p;
	}

	
	public static DatabaseHelper getInstance(Context ctx){
		return new DatabaseHelper(ctx);
	}
}

File: app/src/net/reichholf/dreamdroid/DatabaseHelper.kt
/*
 * © 2010 Stephan Reichholf <stephan at reichholf dot net>
 *
 * Licensed under the Create-Commons Attribution-Noncommercial-Share Alike 3.0 Unported
 * http://creativecommons.org/licenses/by-nc-sa/3.0/
 */
package net.reichholf.dreamdroid

import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.util.Log
import java.io.File
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import net.reichholf.dreamdroid.room.AppDatabase

/**
 * Pre-Room `dreamdroid` SQLite. Read-only profile import for Play 1.x upgrades and
 * cloud-restore snapshots that still have this file. Never creates the database.
 */
object DatabaseHelper {
    const val DATABASE_NAME: String = "dreamdroid"

    val LOG_TAG: String = DatabaseHelper::class.java.simpleName

    private const val PROFILES_TABLE = "profiles"

    private const val KEY_ID = "_id"
    private const val KEY_PROFILE = "profile"
    private const val KEY_HOST = "host"
    private const val KEY_STREAM_HOST = "streamhost"
    private const val KEY_STREAM_PORT = "streamport"
    private const val KEY_FILE_PORT = "fileport"
    private const val KEY_PORT = "port"
    private const val KEY_LOGIN = "login"
    private const val KEY_USER = "user"
    private const val KEY_PASS = "pass"
    private const val KEY_SSL = "ssl"
    private const val KEY_FILE_SSL = "file_ssl"
    private const val KEY_FILE_LOGIN = "file_login"
    private const val KEY_STREAM_LOGIN = "streamlogin"
    private const val KEY_SIMPLE_REMOTE = "simpleremote"
    private const val KEY_DEFAULT_REF = "default_ref"
    private const val KEY_DEFAULT_REF_NAME = "default_ref_name"
    private const val KEY_DEFAULT_REF_2 = "default_ref_2"
    private const val KEY_DEFAULT_REF_2_NAME = "default_ref_2_name"
    private const val KEY_SSID = "ssid"
    private const val KEY_DEFAULT_ON_NO_WIFI = "defaultProfileOnNoWifi"
    private const val KEY_ENCODER_STREAM = "encoder_stream"
    private const val KEY_ENCODER_PATH = "encoder_path"
    private const val KEY_ENCODER_PORT = "encoder_port"
    private const val KEY_ENCODER_LOGIN = "encoder_login"
    private const val KEY_ENCODER_USER = "encoder_user"
    private const val KEY_ENCODER_PASS = "encoder_pass"
    private const val KEY_ENCODER_VIDEO_BITRATE = "encoder_video_bitrate"
    private const val KEY_ENCODER_AUDIO_BITRATE = "encoder_audio_bitrate"
    private const val KEY_TRUST_ALL_CERTS = "trust_all_certs"

    fun databaseFile(context: Context): File = context.getDatabasePath(DATABASE_NAME)

    /**
     * Profiles from an existing leftover file. Empty if the file is missing or
     * unreadable. Does not create `dreamdroid`.
     */
    fun readProfiles(context: Context): List<Profile> {
        val file = databaseFile(context)
        if (!file.exists()) {
            return emptyList()
        }
        return try {
            SQLiteDatabase.openDatabase(
                file.absolutePath,
                null,
                SQLiteDatabase.OPEN_READONLY
            ).use { db ->
                readProfiles(db)
            }
        } catch (e: SQLiteException) {
            Log.e(LOG_TAG, "readProfiles: cannot open leftover $DATABASE_NAME", e)
            emptyList()
        }
    }

    /**
     * When Room has no profiles, copy leftover rows in and delete the file.
     * No-op if Room already has rows or the leftover file is missing.
     * Deletes an existing leftover after a successful open, including 0 rows
     * (empty file created by the old [SQLiteOpenHelper] path).
     */
    fun migrateIntoRoomIfNeeded(
        context: Context,
        dao: Profile.ProfileDao = AppDatabase.profiles(context)
    ): Int = runBlocking(Dispatchers.IO) {
        if (dao.getProfiles().isNotEmpty()) {
            return@runBlocking 0
        }
        val file = databaseFile(context)
        if (!file.exists()) {
            return@runBlocking 0
        }
        val profiles = try {
            SQLiteDatabase.openDatabase(
                file.absolutePath,
                null,
                SQLiteDatabase.OPEN_READONLY
            ).use { db ->
                readProfiles(db)
            }
        } catch (e: SQLiteException) {
            Log.e(LOG_TAG, "migrateIntoRoomIfNeeded: leftover unreadable", e)
            return@runBlocking 0
        }
        for (profile in profiles) {
            profile.id = dao.addProfile(profile).toInt()
        }
        context.deleteDatabase(DATABASE_NAME)
        profiles.size
    }

    private fun readProfiles(db: SQLiteDatabase): List<Profile> {
        val list = ArrayList<Profile>()
        db.query(
            PROFILES_TABLE,
            null,
            null,
            null,
            null,
            null,
            KEY_PROFILE
        ).use { cursor ->
            while (cursor.moveToNext()) {
                list.add(profileFrom(cursor))
            }
        }
        return list
    }

    private fun profileFrom(c: Cursor): Profile {
        var streamPort = c.intOr(KEY_STREAM_PORT, 0)
        if (streamPort <= 0) {
            streamPort = 8001
        }
        var filePort = c.intOr(KEY_FILE_PORT, 0)
        if (filePort <= 0) {
            filePort = 80
        }
        var encoderPath = c.stringOr(KEY_ENCODER_PATH)
        var encoderPort = c.intOr(KEY_ENCODER_PORT, 0)
        var encoderUser = c.stringOr(KEY_ENCODER_USER)
        var encoderPass = c.stringOr(KEY_ENCODER_PASS)
        var encoderAudioBitrate = c.intOr(KEY_ENCODER_AUDIO_BITRATE, 0)
        var encoderVideoBitrate = c.intOr(KEY_ENCODER_VIDEO_BITRATE, 0)
        encoderPath = encoderPath ?: "stream"
        encoderPort = if (encoderPort <= 0) 554 else encoderPort
        encoderUser = encoderUser ?: ""
        encoderPass = encoderPass ?: ""
        encoderAudioBitrate = if (encoderAudioBitrate <= 0) 128 else encoderAudioBitrate
        encoderVideoBitrate = if (encoderVideoBitrate <= 0) 2500 else encoderVideoBitrate

        val profile = Profile(
            c.intOr(KEY_ID, 0).takeIf { it > 0 },
            c.stringOr(KEY_PROFILE),
            c.stringOr(KEY_HOST),
            c.stringOr(KEY_STREAM_HOST),
            c.intOr(KEY_PORT, 0),
            streamPort,
            filePort,
            c.boolOr(KEY_LOGIN),
            c.stringOr(KEY_USER),
            c.stringOr(KEY_PASS),
            c.boolOr(KEY_SSL),
            c.boolOr(KEY_TRUST_ALL_CERTS),
            c.boolOr(KEY_STREAM_LOGIN),
            c.boolOr(KEY_FILE_LOGIN),
            c.boolOr(KEY_FILE_SSL),
            c.boolOr(KEY_SIMPLE_REMOTE),
            c.stringOr(KEY_DEFAULT_REF),
            c.stringOr(KEY_DEFAULT_REF_NAME),
            c.stringOr(KEY_DEFAULT_REF_2),
            c.stringOr(KEY_DEFAULT_REF_2_NAME),
            c.boolOr(KEY_ENCODER_STREAM),
            encoderPath,
            encoderPort,
            c.boolOr(KEY_ENCODER_LOGIN),
            encoderUser,
            encoderPass,
            encoderVideoBitrate,
            encoderAudioBitrate
        )
        profile.ssid = c.stringOr(KEY_SSID)
        profile.isDefaultProfileOnNoWifi = c.boolOr(KEY_DEFAULT_ON_NO_WIFI)
        return profile
    }

    private fun Cursor.intOr(name: String, default: Int): Int {
        val index = getColumnIndex(name)
        if (index < 0 || isNull(index)) {
            return default
        }
        return getInt(index)
    }

    private fun Cursor.stringOr(name: String): String? {
        val index = getColumnIndex(name)
        if (index < 0 || isNull(index)) {
            return null
        }
        return getString(index)
    }

    private fun Cursor.boolOr(name: String): Boolean = intOr(name, 0) == 1
}
